Update jerry API

LaszloLango opened the pull request:
* Fix error handling (related issue: #1141)
 * Move output paramters to the end of the arguments lists

JerryScript-DCO-1.0-Signed-off-by: László Langó llango.u-szeged@partner.samsung.com

zherczeg commented on jerry-core/jerry.c:
```diff
@@ -125,9 +125,18 @@ jerry_make_api_unavailable (void)
  * Returns whether the given jerry_value_t is null.
  */
-bool
+inline bool __attr_always_inline___
```
I think this depends on LTO/all-in-one build, and an API user unlikely to have this advantage.

LaszloLango replied:
You are right. This was not a good idea. I'll remove it.

zherczeg commented on jerry-core/jerry-api.h:
```diff
@@ -197,15 +197,15 @@ bool jerry_is_constructor (const jerry_object_t *);
-bool jerry_construct_object (jerry_object_t *, jerry_value_t *, const jerry_value_t[], uint16_t);
-bool jerry_call_function (jerry_object_t *, jerry_object_t *, jerry_value_t *, const jerry_value_t[], uint16_t);
+bool jerry_construct_object (jerry_object_t *, const jerry_value_t[], uint16_t, jerry_value_t *);
+bool jerry_call_function (jerry_object_t *, jerry_object_t *, const jerry_value_t[], uint16_t, jerry_value_t *);
```
Why don't these functions return with jerry_value_t?

LaszloLango replied:
Was not sure about that, but I don't mind to change.

LaszloLango commented:
@zherczeg, I've upadted the PR. Please check.

zherczeg commented on jerry-core/jerry.c:
```diff
@@ -1098,11 +1100,9 @@ jerry_set_object_native_handle (jerry_object_t *object_p, /**< object to set han
- * @return true, if invocation was performed successfully, i.e.:
- *                - no unhandled exceptions were thrown in connection with the call;
- *         false - otherwise.
+ * @return TODO
```
return todo?

LaszloLango replied:
sorry, I forgot to update the description.

zherczeg commented on jerry-core/jerry.c:
```diff
@@ -1157,29 +1152,19 @@ jerry_invoke_function (bool is_invoke_as_constructor, /**< true - invoke functio
-jerry_construct_type_error (jerry_value_t *retval_p) /**< [out] value with constructed
-                                                      *         TypeError object */
+static inline jerry_value_t __attr_always_inline___
+jerry_construct_type_error (void)
```
inline again. I think inlining is important, but it is unclear to me how can we take advantage of that, so I would do it as a separate step. Such change would not even hurt the api.

LaszloLango replied:
This is a static function. It cannot be used outside of `jerry.c`. I think this won't be a problem.

zherczeg commented on jerry-core/jerry.c:
```diff
@@ -1189,20 +1174,12 @@ jerry_construct_type_error (jerry_value_t *retval_p) /**< [out] value with const
- *         false - otherwise, 'retval_p' contains thrown exception:
- *                  if called object is not function object - a TypeError instance;
- *                  else - exception, thrown during the function call.
+ * @return TODO
```
Fix all ot these please.

zherczeg commented:
The question is what should happen if we pass error objects as arguments is still open. We definitely need to do something with that before the release.

zherczeg commented:
LGTM

dbatyai commented on main-unix.c:
```diff
@@ -461,12 +462,11 @@ main (int argc,
+                                                           NULL,
+                                                           args,
+                                                           1);
+        if (!jerry_value_is_error (ret_val_print))
```
`ret_val_print` should also be released when it's an error, otherwise it may leak.

dbatyai commented on jerry-core/jerry.c:
```diff
@@ -946,35 +954,29 @@ jerry_foreach_object_field (jerry_object_t *object_p, /**< object */
- * @return true, if field value was retrieved successfully, i.e. upon the call:
- *                - there is field with specified name in the object;
- *         false - otherwise.
+ * @return ecma value
```
In some places it's `ecma value` and in other places it's `jerry value`. I think the naming should be consistent.
Also, these `@return`s could be a little bit more descriptive in my opinion. I get that this is a simple getter function, but still.

LaszloLango commented:
@dbatyai, I've updated the PR. I will do a big review and update on the function comments in a later PR, but the comments of touched functions should be ready (more or less), so I hope I fixed all of them. :)

dbatyai commented:
LGTM
